xbmc/video/PlayerController.cpp

```diff
@@ -328,6 +328,61 @@ bool CPlayerController::OnAction(const CAction &action)
         return true;
       }
 
+      case ACTION_VSHIFT_TOP:
+      {
+        CVideoSettings vs = g_application.GetAppPlayer().GetVideoSettings();
+        vs.m_CustomVerticalShift = -1.0f; // -2.0f > 2.0f
+        vs.m_ViewMode = ViewModeCustom;
+        g_application.GetAppPlayer().SetRenderViewMode(ViewModeCustom, vs.m_CustomZoomAmount,
+                                                   vs.m_CustomPixelRatio, vs.m_CustomVerticalShift,
+                                                   vs.m_CustomNonLinStretch);
+        return true;
+      }
+
+      case ACTION_VSHIFT_CENTER:
+      {
+        CVideoSettings vs = g_application.GetAppPlayer().GetVideoSettings();
+        vs.m_CustomVerticalShift = 0.0f; // -2.0f > 2.0f
+        vs.m_ViewMode = ViewModeCustom;
+        g_application.GetAppPlayer().SetRenderViewMode(ViewModeCustom, vs.m_CustomZoomAmount,
+                                                   vs.m_CustomPixelRatio, vs.m_CustomVerticalShift,
+                                                   vs.m_CustomNonLinStretch);
+        return true;
+      }
+
+      case ACTION_VSHIFT_BOTTOM:
+      {
+        CVideoSettings vs = g_application.GetAppPlayer().GetVideoSettings();
+        vs.m_CustomVerticalShift = 1.0f; // -2.0f > 2.0f
+        vs.m_ViewMode = ViewModeCustom;
+        g_application.GetAppPlayer().SetRenderViewMode(ViewModeCustom, vs.m_CustomZoomAmount,
+                                                   vs.m_CustomPixelRatio, vs.m_CustomVerticalShift,
+                                                   vs.m_CustomNonLinStretch);
+        return true;
+      }
+
+      case ACTION_VSHIFT_TOP239:
+      {
+        CVideoSettings vs = g_application.GetAppPlayer().GetVideoSettings();
+        vs.m_CustomVerticalShift = -1.12f; // -2.0f > 2.0f
+        vs.m_ViewMode = ViewModeCustom;
+        g_application.GetAppPlayer().SetRenderViewMode(ViewModeCustom, vs.m_CustomZoomAmount,
+                                                   vs.m_CustomPixelRatio, vs.m_CustomVerticalShift,
+                                                   vs.m_CustomNonLinStretch);
+        return true;
+      }
+
+      case ACTION_VSHIFT_BOTTOM239:
+      {
+        CVideoSettings vs = g_application.GetAppPlayer().GetVideoSettings();
+        vs.m_CustomVerticalShift = 1.12f; // -2.0f > 2.0f
+        vs.m_ViewMode = ViewModeCustom;
+        g_application.GetAppPlayer().SetRenderViewMode(ViewModeCustom, vs.m_CustomZoomAmount,
+                                                   vs.m_CustomPixelRatio, vs.m_CustomVerticalShift,
+                                                   vs.m_CustomNonLinStretch);
+        return true;
```

Comment:
I see lots of code copies here. Please factor out these copies into a new method which takes the new value for `vs.m_CustomVerticalShift` as a parameter and call the new method from the different case labels.